sm/json_sm/src/messages.rs

```diff
@@ -0,0 +1,384 @@
+    pub fn new(req: &SoftwareListRequest) -> SoftwareListResponse {
+        let response = SoftwareRequestResponse {
+            id: req.id,
+            status: SoftwareOperationStatus::Successful,
+            reason: None,
+            current_software_list: vec![],
+            failures: vec![],
+        };
+
+        SoftwareListResponse {
+            response
+        }
+    }
+
+    pub fn add_modules(&mut self, plugin_type: &str, modules: Vec<SoftwareModule>) {
+        self.response.current_software_list.push(SoftwareRequestResponseSoftwareList {
+            plugin_type: plugin_type.to_string(),
+            modules: modules.into_iter().map(|module| module.into()).collect::<Vec<SoftwareModuleItem>>(),
+        })
+    }
+
+    pub fn set_error(&mut self, reason: &str) {
+        self.response.status = SoftwareOperationStatus::Failed;
+        self.response.reason = Some(reason.into());
+    }
+
+    pub fn id(&self) -> usize {
+        self.response.id
+    }
+
+    pub fn error(&self) -> Option<String> {
+        self.response.reason.clone()
+    }
+}
+
+/// Variants represent Software Operations Supported actions.
+#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum SoftwareModuleAction {
+    Install,
+    Remove,
+}
+
+/// Software module payload definition.
+#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct SoftwareModuleItem {
+    pub name: SoftwareName,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub version: Option<SoftwareVersion>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub url: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub action: Option<SoftwareModuleAction>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub reason: Option<String>,
+}
+
+/// Software Operation Response payload format.
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct SoftwareRequestResponse {
+    // TODO: Is this the right approach, maybe nanoid?
+    pub id: usize,
+    pub status: SoftwareOperationStatus,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub reason: Option<String>,
+
+    #[serde(default)]
+    pub current_software_list: Vec<SoftwareRequestResponseSoftwareList>,
+
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub failures: Vec<SoftwareRequestResponseSoftwareList>,
+}
+
+impl<'a> Jsonify<'a> for SoftwareRequestResponse {}
+
+impl SoftwareRequestResponse {
+    pub fn new(id: usize, status: SoftwareOperationStatus) -> Self {
+        SoftwareRequestResponse {
+            id,
+            status,
+            current_software_list: vec![],
+            reason: None,
+            failures: vec![],
+        }
+    }
+
+    pub fn finalize_response(&mut self, software_list: Vec<SoftwareRequestResponseSoftwareList>) {
```

Comment:
I suppose it is a leftover from my code but I suppose `software_list` should be a ref to Vec and not Vec, right?
Or maybe even better a slice?
```suggestion
    pub fn finalize_response(&mut self, software_list: &[SoftwareRequestResponseSoftwareList]) {
```